member_map.go

```diff
@@ -198,36 +170,56 @@ func override(newMem Member, existingMem Member) Member {
 	return existingMem
 }
 
-// Update member and update waitingMembers
-// todo
-func (m *MemberMap) UpdateMember(member Member) error {
-	return nil
-}
+// Process Alive message
+// Return 2 parameter bool, error bool means Changed in MemberList
+// 1. If aliveMessage Id is empty return false and ErrEmptyMemberID
+// 2. if aliveMessage Id is not in MemberList then Create Member and Add MemberList
+// 3. if aliveMessage Id is in MemberList and existingMember's Incarnation is bigger than AliveMessage's Incarnation Than return false, ErrIncarnation
+// 4. if aliveMessage Id is in MemberList and AliveMessage's Incarnation is bigger than existingMember's Incarnation Than Update Member and Return true, nil
+func (m *MemberMap) Alive(aliveMessage AliveMessage) (bool, error) {
 
-// Remove member and update waitingMembers
-// todo
-func (m *MemberMap) RemoveMember(member Member) error {
-	return nil
-}
+	m.lock.Lock()
+	defer m.lock.Unlock()
 
-// This function will be called when memberMap updated
-// Create waiting memberID List and shuffle
-func resetWaitingMembersID(memberMap map[MemberID]Member) []Member {
+	// Check whether Member Id empty
+	if aliveMessage.ID == "" {
+		return false, ErrEmptyMemberID
+	}
+	// Check whether it is already exist
+	existingMem, ok := m.members[MemberID{aliveMessage.ID}]
 
-	// Convert Map to List
-	waitingMembersID := make([]Member, 0)
-	for _, member := range memberMap {
-		waitingMembersID = append(waitingMembersID, member)
+	// if Member is not exist in MemberList
+	if !ok {
+		m.members[MemberID{aliveMessage.ID}] = createMember(aliveMessage.MemberMessage, Alive)
+		return true, nil
 	}
+	// Check incarnation
+	if aliveMessage.Incarnation <= existingMem.Incarnation {
+		return false, ErrIncarnation
+	}
```

Comment:
In my opinion, receiving `aliveMessage.Incarnation <= existingMem.Incarnation` is not an error. it's just old message. how about just logging?